Python/pytime.c

```diff
@@ -781,35 +869,71 @@ _PyTime_GetMonotonicClockWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 
 
 #ifdef MS_WINDOWS
+static int
+get_perf_frequency(LONGLONG *t0, LONGLONG *perf_frequency,
+                   _PyTime_t *num_p, _PyTime_t *den_p)
+{
+    LARGE_INTEGER now;
+    LARGE_INTEGER freq;
+    _PyTime_t num, den, gcd;
+
+    QueryPerformanceCounter(&now);
+    if (!QueryPerformanceFrequency(&freq)) {
+        PyErr_SetFromWindowsErr(0);
+        return -1;
+    }
+    if (freq.QuadPart < 1) {
+        return -1;
+    }
+
+    *t0 = now.QuadPart;
+    *perf_frequency = freq.QuadPart;
+
+    /* Compute the greatest common divisor to minimize numerator and
+       denominator, to reduce the risk of integer overflow in
+       _PyTime_GetWinPerfCounterWithInfo(). */
+    num = SEC_TO_NS;
+    den = freq.QuadPart;
+    gcd = _PyTime_GCD(num, den);
+    num /= gcd;
+    den /= gcd;
+    if (num < 1 || den < 1) {
+        return -1;
+    }
+
+    *num_p = num;
+    *den_p = den;
+    return 0;
+}
+
 int
-_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *t, _Py_clock_info_t *info)
+_PyTime_GetWinPerfCounterWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
 {
-    static LONGLONG cpu_frequency = 0;
-    static LONGLONG ctrStart;
+    static LONGLONG perf_frequency = 0;
+    static _PyTime_t numer = 0, denom = 0;
+    static LONGLONG t0 = 0;
     LARGE_INTEGER now;
-    double diff;
+    _PyTime_t t;
 
-    if (cpu_frequency == 0) {
-        LARGE_INTEGER freq;
-        QueryPerformanceCounter(&now);
-        ctrStart = now.QuadPart;
-        if (!QueryPerformanceFrequency(&freq) || freq.QuadPart == 0) {
-            PyErr_SetFromWindowsErr(0);
+    if (perf_frequency == 0) {
+        if (get_perf_frequency(&t0, &perf_frequency, &numer, &denom) < 0) {
```

Comment:
I almost missed that this initialised `t0` in addition to `perf_frequency`. Since it initialises all the statics for this function rather than just `perf_frequency`, perhaps `_init_win_perf_counter` would be a better name?

Reply:
I renamed the method to init_win_perf_counter(). No need for a "_" prefix, since the function is marked as private using "static" keyword.